# Python Data Structures

-----

In previous lessons, we introduced the basic Python programming concepts that allow you to write simple programs. In this notebook, we will introduce four powerful, built-in data structures that provide an alternate approach to organizing data. These four data structures are the string, tuple, list, and dictionary. Learning how to work effectively with these data structures will allow you to develop concise, fast applications.

## Table of Contents

[Python Data Structures](#Python-Data-Structures)  

[String](#String)  

[List](#List)  

[Dictionary](#Dictionary)  

[Tuple](#Tuple)  

[Slicing](#Slicing)  

[Common Sequence Operations](#Common-Sequence-Operations)


-----
[[Back to TOC]](#Table-of-Contents)

## Python Data Structures

Python provides built-in support for a number of useful data structures, including the four introduced in this notebook: string, tuple, list, and dictionary. In this section, we briefly introduce these four data structures and provide an example of each.

__String__: 
A sequence of zero or more characters that are enclosed within either a pair of single quote characters, `'`, or a pair of double quote characters, `"`. A Python string is an instance of class `str`.

`'A string containing many characters'`

__Tuple__:
An ordered sequence of zero or more values that are enclosed in parentheses, `(` and `)`. The different values in a tuple are generally separated by commas. A Python tuple is an instance of class `tuple`.

`(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)`

__List__:
An ordered sequence of zero or more values that are enclosed in square brackets, `[` and `]`. The different values in the list are separated by commas. A Python list is an instance of class `list`.

`[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]`

__Dictionary__:
An **unordered** collection of key-value pairs that are enclosed in curly braces, `{` and `}`. A key is separated from its corresponding value by a colon `:` and the different key-value entries in the collection are separated by commas. A Python dictionary is an instance of class `dict`.

`d = {'name': "Alexander", 'age': 30, 'location': (102.1, 32.1)}`

----

Of these four data structures, the _list_ and the _dictionary_ are **mutable**, which means the values stored in these structures can be changed. On the other hand,  the _string_ and the _tuple_ are **immutable**, which means the values cannot be changed once created. Instead, a new data structure must be created, either explicitly by the programmer, or implicitly by the Python interpreter. 

All of these data structures can be displayed by using the Python built-in `print` function, which converts its arguments to a string, which is then sent to STDOUT (which in this course is generally the space in the notebook immediately following the `print` function):

```python
data = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
print(data)
```

The _dictionary_ is a special data structure that maps keys to values, and thus the keys, which can be stored in any order, are used to access the values. The other three structures are ordered sequences, and thus individual elements can be accessed by specifying an index position within square brackets, `[]`, with the caveat that Python is a zero-indexed language. Thus, given an ordered sequence `data`, the following accesses are legal:

`data = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]`
- `data[0]`: access the first value which is `1`
- `data[1]`: access the second value which is `2`
- `data[-1]`: access the last value which is `10`
- `data[-2]`: access the second to last value which is `9`

-----

-----
[[Back to TOC]](#Table-of-Contents)


## String

In Python, a string is a sequence of zero or more characters that are enclosed within either a pair of single quote characters, `'`, or a pair of double quote characters, `"`. While it might seem confusing to have two such similar representations, it can be quite handy, as this allows strings to be written that include either single quotes or double quote characters by simply using the opposite type to enclose the string itself:

```
"A string with a contraction such as don't or a possessive like Python's string."
'A string with a quote, "Four score and seven years ago ..."'
```

A string can span multiple lines by using either three single or double quote characters to enclose the sequence of characters, just like multi-line comments.

A Python string object has many associated methods that make this data structure even more useful, like changing the case of the letters or searching for sub-strings. These methods will be introduced in a subsequent lesson. In the following Code cell, we demonstrate basic string operations.

----

In [1]:
text = 'The brown dog jumped over the quick fox!'

# Add string manipulations

print(text)

print(text[0:10])

print(text[-4:])

print('Text length = ', len(text))
print('Upper case:', text.upper())
print('Lower case:', text.lower())

The brown dog jumped over the quick fox!
The brown 
fox!
Text length =  40
Upper case: THE BROWN DOG JUMPED OVER THE QUICK FOX!
Lower case: the brown dog jumped over the quick fox!


-----

<font color='red' size = '5'> Student Exercise </font>

In the empty **Code** cell below, first create a new string called 'mystring' that contains at least 12 characters (e.g., `mystring = 'This is a demo string.'`). Next, write two separate `print` functions that display the third through sixth characters and the last character from the `mystring` string.

-----

-----

### String Modification

At the start of this notebook, we introduced the string as an immutable sequence of characters. Fundamentally, this means we can't directly change (or mutate) an existing string. Instead, to modify an existing string, we first must create a new string, and second, copy the old string and new string into the appropriate place. This is demonstrated in a straightforward manner in the following Code cell. This approach is not optimal; we will see faster approaches to construct large strings in a future lesson. In addition, we demonstrate that the `text` variable points to a new string by checking the variable's id, which changes after this string modification process.

-----

In [2]:
# Create test string and show text and variable id
text = "Hello"
print(text)
print(id(text))

# Print blank line
print()

# Try to modify string, but notice new variable id
text += ' World!'
print(text)
print(id(text))

Hello
4597095256

Hello World!
4599008688


---

### Special String Types

Python provides two special types of string literals: the _raw string literal_ and the _format string literal_ (there are also byte literals and unicode literals, but we will not discuss them in this course). A raw string is not processed by the Python interpreter; the content of the string is left alone. For example, backslashes are not handled in a special way when processing the string. To create a raw string, simply prefix the string with either a lowercase `'r'` or an uppercase `'R'`:

```python
data = r'Text with a \ that is left unprocessed.'
```

The difference between a normal and raw string is demonstrated in the following Code cell, where the normal string converts the `'\n'` character sequence into a newline, while the raw string leaves the character sequence alone. This is useful when constructing regular expressions or other strings that might include formatting information, like plot axes that use LaTeX formatting.

-----

In [3]:
# Make and print same text as raw and normal strings
r_text = r'Hello World!\nHere I am!'
text = 'Hello World!\nHere I am!'

print(r_text)
print(text)

Hello World!\nHere I am!
Hello World!
Here I am!


-----

The second special type of string is the [_formatted string literal_ (or f-string)][pfs]. The format string is used to dynamically create a new string by using the values of other variables. This technique is often used to generate dynamic, formatted output. To create an f-string, simply prefix the regular string with either a lowercase `'f'` or an uppercase `'F'`:

```python
x = 10
data = f'The value of x = {x}.'
```

The Python interpreter will replace the character sequence `{x}` with the character sequence `10`, which is the value of the variable `x`. This is demonstrated in the following Code cell, where the dynamically created string is displayed.

-----
[pfs]: https://docs.python.org/3/reference/lexical_analysis.html#f-strings

In [4]:
# Demonstrate f-string creation and use.

a = 2; b = 4; c = 8

print(f'a = {a}, b = {b}, and c = {c}')

a = 2, b = 4, and c = 8


-----

In earlier versions of Python, dynamic string formatting was done using the `format` function on a string. Both techniques work nearly the same way, with the variable to be replaced enclosed in curly braces inside the string. However, the f-string approach is cleaner and easier to write (since everything is contained within the string itself), so we will focus solely on the newer approach.

The benefit of using an f-string is that the exact format for the conversion of the value of a variable to a string can be defined. This includes the data type, the width (or number of characters to use), the precision for a decimal number, and whether padding (e.g., by using zeros) should be used. While the rules can appear complex, actually using f-strings can be simple, as demonstrated in the following Code cell.

-----

In [5]:
a = 2; b = 4; c = 8

print(f'a = {a:04}, b = {b:4}, and c = {c:4.2f}')

a = 0002, b =    4, and c = 8.00


In [6]:
a = 'Alexander' ; b = 6.43 ; c = 4 / 2.3

print(f'a = {a:10s}, b = {b:7.3f}, and c = {c:08.4f}')

a = Alexander , b =   6.430, and c = 001.7391


-----

In the first example, the same data that were displayed in the original f-string example are used to create a new string by using a revised f-string syntax. First, the value of the `a` variable is displayed in a width of four characters with zeros padding any extra space. This means we insert `0002` into the new string. Second, the value of the variable `b` is added to the string with a width of four characters but using the default space padding, which means we insert `   4` into the new string. This can be used to align columns when displaying data. Finally, the value of the `c` variable is displayed as a floating-point value (indicated by the `f` character at the end of the sequence). In addition, we specify a width of four characters with two of the characters to appear after the decimal point, which means we insert ` 8.00` into the new string.

In the second Code cell, several other format specifiers are demonstrated. First, the value of the variable `a` is converted to a string of width ten characters and inserted into the new string. Next, the value of the variable `b` is converted to a floating-point literal that is seven characters wide with four characters after the decimal point, which means `  6.430` is inserted into the new string (note this is two spaces at the front, because the decimal point is included in the total of seven characters). Finally, the value of the variable `c` is converted to a floating-point number, padded with leading zeros, that is eight characters wide with four numbers after the decimal point.

-----

<font color='red' size = '5'> Student Exercise </font>

In the empty **Code** cell below, first create three variables (you can pick their name) to hold the values 0, 3.14, and 'Hello World!'. Now create a new f-string to hold these values. First, pad with spaces the value of the first variable to be width 10. Second, convert the second variable to be of width 10 with 3 characters after the decimal point and pad with leading zeros. Finally, convert the final variable to be of width 20 with padding spaces. Display the new string using the print statement.

-----

-----
[[Back to TOC]](#Table-of-Contents)


## List

A list is a mutable sequence that can hold homogeneous data, `[1, 2, 3, 4, 5]` or heterogeneous data, `[1, '2', 'Three', (4, 5)]`. Lists are very powerful data structures and are used extensively in many Python programs. A list can be created in several different ways:

1. `[]`: An empty list
2. `[1]`: A single valued list
3. `[1, 2, 3]`: Comma-separated items
4. `list()`: Using the `list` class constructor

Since a `list` is mutable, a list can be changed by adding elements, removing elements, or simply changing existing elements in place.

By default, assigning a list to a new variable results in a shallow copy, which means that both variables point to the same underlying list and any changes to one results in changes to the other. For example, after this set of operations:

```python
data = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
d = data
d[0] = -1
```

In the shallow copy we simply made a variable that pointed at the original version of the data list, so `data[0]` now contains the value `-1`. To obtain a deep copy, use the slice notation without any values. For example, after this set of operations:

```python
data = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
d = data[:]
d[0] = -1
```

In this case, `data[0]` retains the original value of `1` because the modification changed the new value, not the original value.

In the following Code cells, we first demonstrate how to slice elements of a list. Next, we apply several built-in functions to a list. Finally, we demonstrate a deep and shallow copy, where modifying a deep copy does not change the original list, while modifying a shallow copy does modify the original list.

-----

In [7]:
# Create a list and display the result.
data = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

print('Original', data)

print('Sliced', data[-4:])

Original [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Sliced [7, 8, 9, 10]


In [8]:
# Apply built-in functions to the list
print('Data length = ', len(data))
print('Min value = ', min(data))
print('Max value = ', max(data))

Data length =  10
Min value =  1
Max value =  10


In [9]:
# Now we compare shallow and deep copies
print('Original', data)

# First a deep copy and try to modify
d = data[:]
d[1] = -1

print('Deep Copy', data)

# Now a shallow copy and try to modiy
d = data
d[1] = -1

print('Shallow Copy', data)

Original [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Deep Copy [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Shallow Copy [1, -1, 3, 4, 5, 6, 7, 8, 9, 10]


-----

<font color='red' size = '5'> Student Exercise </font>

In the empty **Code** cell below, first create a new list called 'mylist' that holds at least 10 values. Next, write three separate `print` functions that display the third element, sixth element, and the last element from the `mylist` list.

-----

-----

### List Modification

Earlier, we defined a list as a mutable collection of objects (or other data items). This means a list can be directly modified once created. This flexibility means that a `list` will likely be slower than comparable data structures, such as a NumPy array, which we will learn about in a future lesson, or even a `tuple`. However, this flexibility makes lists very powerful and extremely useful. 

The following Code cells first demonstrate how to modify a list in place before demonstrating how to add elements to a list via the list `append` method. Notice the difference between adding a new list element one at a time and adding all elements as a list (the latter adds a new list element that consists of a new list). Additional list methods will be introduced in a subsequent lesson. 

When modifying a list, we also show the list variable id to demonstrate that the same list is being modified.

-----

In [10]:
# Modify odd elements

print('Original', data)
print(id(data))

data[1] = -1 * data[1] 
data[3] = -1 * data[3] 
data[5] = -1 * data[5] 
data[7] = -1 * data[7] 
data[9] = -1 * data[9] 

print('Modified', data)
print(id(data))

Original [1, -1, 3, 4, 5, 6, 7, 8, 9, 10]
4597051080
Modified [1, 1, 3, -4, 5, -6, 7, -8, 9, -10]
4597051080


In [11]:
# First copy original data list
dc_data = data[:]
print('Original', data)
print(id(data))

# Add list elements one at a time
data.append(11)
data.append(12)
data.append(13)
data.append(14)
data.append(15)
data.append(16)
data.append(17)
data.append(18)
data.append(19)
data.append(20)

print('Appended', data)
print(id(data))

# Try adding list en masse
dc_data.append([11, 12, 13, 14, 15, 16, 17, 18, 19, 20])
print('Bulk Appended', dc_data)

Original [1, 1, 3, -4, 5, -6, 7, -8, 9, -10]
4597051080
Appended [1, 1, 3, -4, 5, -6, 7, -8, 9, -10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]
4597051080
Bulk Appended [1, 1, 3, -4, 5, -6, 7, -8, 9, -10, [11, 12, 13, 14, 15, 16, 17, 18, 19, 20]]


-----
[[Back to TOC]](#Table-of-Contents)


## Dictionary

A dictionary is an **unordered** sequence that can hold values that are referenced by corresponding keys. These key-value pairs are separated by commas, while the value is separated from its key by a colon ':' character. A dictionary can be created in several different ways:

1. `{}`: An empty dictionary
2. `{'1': 1}`: A single key-value dictionary
3. `{'1': 1, '2': "two", '3': (1, 2, 3)}`: Comma-separated key-value pairs
4. `dict()`: Using the `dict` class constructor

Since a `dictionary` is mutable, a dictionary can be changed by adding key-value pairs, by removing key-value pairs, or by simply changing existing values in place. We demonstrate the last approach in the following Code cells; the other capabilities are provided by methods that will be introduced in a subsequent lesson.

In the following Code cells, we create a simple dictionary and apply different operations that demonstrate how to use simple dictionaries.

-----

In [12]:
# Create dictionary
d = {'1': 1, '2': 'Two', '3': [1, 2, 3]}

# Display Dictionary and basic info.
print('Original Dictionary', d)
print('Dictionary id', id(d))

# Do basic membership tests
print('Test membership of key "1":', '1' in d)
print('Test absence of key "4":', '4' not in d)

# Modify an value given the key
d['2'] = 'Three'
print('Modified Dictionary', d)
print('Dictionary id', id(d))

Original Dictionary {'1': 1, '2': 'Two', '3': [1, 2, 3]}
Dictionary id 4599211568
Test membership of key "1": True
Test absence of key "4": True
Modified Dictionary {'1': 1, '2': 'Three', '3': [1, 2, 3]}
Dictionary id 4599211568


In [13]:
# Basic dictionary operations
print('Dictionary length = ', len(d))
print('Keys:', d.keys())
print('Values:', d.values())

print('Value of key "3":', d['3'])

Dictionary length =  3
Keys: dict_keys(['1', '2', '3'])
Values: dict_values([1, 'Three', [1, 2, 3]])
Value of key "3": [1, 2, 3]


-----

<font color='red' size = '5'> Student Exercise </font>

In the empty **Code** cell below, first create a new dictionary called 'mydict' that contains at least five key-value pairs (e.g., `mydict = {'one' : 1, 'two' : 2, 'three' : 3, 'four' : 4, 'five' : 5}`). Next, write two separate `print` functions that display the value of the key `two` and the key `five` from the `mydict` dictionary.

-----

-----
[[Back to TOC]](#Table-of-Contents)


## Tuple

A tuple is an **immutable** sequence that can hold homogeneous data, `(1, 2, 3, 4, 5)` or heterogeneous data, `(1, '2', 'Three', (4, 5))`. A tuple can be created in several different ways:

1. `()`: An empty tuple
2. `1,` or `(1, )`: A single valued tuple
3. `1, 2, 3` or `(1, 2, 3)`: Comma-separated items
4. `tuple()`: Using the `tuple` class constructor

Note the requirement for the trailing comma to create a single-valued tuple; otherwise the Python interpreter interprets the expression as a value enclosed in parentheses; for example `(1)` is an integer. Any change to a `tuple` requires the creation of a new `tuple`.

Tuples are commonly used to pass information to and from functions and allow for the assignment of multiple data values simultaneously via _unpacking_. A function with more than one return values returns values in a tuple as shown in the following code cells.

-----

In [14]:
def divide(dividend, divisor):
    """
    Calculate dividend divided by divisor
    Return: quotient and remainder
    """

    quotient = dividend // divisor
    remainder = dividend % divisor
    return quotient, remainder

In [15]:
divide(20, 3)

(6, 2)

In [16]:
type(divide(20, 3))

tuple

In [17]:
#unpack tuple to variables
q, r = divide(20, 3)
print (f'quotient is {q}, remainder is {r}.')

quotient is 6, remainder is 2.


-----

<font color='red' size = '5'> Student Exercise </font>

In the empty **Code** cell below, first create a new tuple called 'mytup' that contains at least 10 values.  Next, write two separate `print` functions that display the value of the fourth element and the second to last element from the `mytup` tuple.

-----

-----
[[Back to TOC]](#Table-of-Contents)

## Slicing

Python supports a rich array of techniques for extracting values from an ordered list beyond the single value access method, known as slicing. Given an ordered sequence `data`, the basic format is `data[start:end:stride]` where start and end are the starting and ending index values, respectively, and stride is the number of values to skip when iterating. If start or end are omitted, the default is the first and last value, while the default stride is one. A negative value can be used for either the start or the end index values, which indicates relative to the end value. In `data[start:end:stride]`, `start` has default value 0, `end` has default value which is the length of the object, `stride` has default value `1`. If a value is not provided, default value will be used. So `data[::]` or `data[:]` means *slice data from beginning to end with stride 1*, which has same data as `data`. When `start` or `end` is negative, it means count from behind. `stride` can also be negative which means traverse backwardly. Starting index value is included and ending index value is **not** included in the slicing result. These concepts are demonstrated in the following code block:

In [18]:
# Edit the start/end/stride values to learn slicing

data = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

print(data[0])
print(data[2:-2])
print(data[:-3:2])

#get odd numbers in the list
print('Odd numbers:', data[::2])

#reverse list
print('Reversed list:', data[-1::-1])

1
[3, 4, 5, 6, 7, 8]
[1, 3, 5, 7]
Odd numbers: [1, 3, 5, 7, 9]
Reversed list: [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]


In [19]:
# Now slice a tuple
t_data = (1, 2, 3, 4, 5, 6, 7, 8, 9, 10)

print(t_data[0])
print(t_data[2:-2])
print(t_data[:-3:2])

#get even numbers in the tuple
print('Even numbers:', t_data[1::2])

1
(3, 4, 5, 6, 7, 8)
(1, 3, 5, 7)
Even numbers: (2, 4, 6, 8, 10)


In [20]:
# Now slice a string
s_data = '1,2,3,4,5,6,7,8,9,10'

print(s_data[0])
print(s_data[2:-2])
print(s_data[:-3:2])

1
2,3,4,5,6,7,8,9,
123456789


-----

<font color='red' size = '5'> Student Exercise </font>

In the empty **Code** cell below, first create a list that has the elements zero through fifteen, inclusive. Next, use slicing to print out the first element, last element, odd elements, and every fourth element

-----

-----
[[Back to TOC]](#Table-of-Contents)


## Common Sequence Operations

In addition to slicing, the string, tuple, and list support several common sequence operations. Given a value `v`, integer `n`, and similar typed sequences `s` and `t`:

| Operation         | Description                              |
| ----------------- | ---------------------------------------- |
| `v in s`          | `True` if `v` is in the sequence `s`, otherwise `False` |
| `v not in s`      | `False` if `v` is in the sequence `s`, otherwise `True` |
| `s + t`           | concatenation of `s` and `t`             |
| `s * n` or `n* s` | `n` shallow copies of `s` concatenated   |
| `len(s)`          | the number of elements in the sequence `s` |
| `min(s)`          | the smallest elements in the sequence `s` |
| `max(s)`          | the largest of elements in the sequence `s` |
| `s.count(v)`      | number of times `v` appears in `s`       |

These methods are demonstrated below on the `data` list:

In [21]:
data = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
1 in data

True

In [22]:
0 not in data

True

In [23]:
data * 2

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [24]:
data + [11, 12, 13, 14]

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]

In [25]:
print(len(data), min(data), max(data))

10 1 10


## Ancillary Information

The following links are to additional documentation that you might find helpful in learning this material. Reading these web-accessible documents is completely optional.

1. The official Python documentation for [strings][1a], [lists][1b], [dictionaries][1c], and [tuples][1d]
2. The book [_A Byte of Python_](https://python.swaroopch.com/data_structures.html) introduces these data structures.
3. A discussion on the [_native_ data types][2] mentioned in this notebook from the book, _Dive into Python_
4. The book [_Think Python_][3] includes a discussion on these data structures.

-----

[1a]: https://docs.python.org/3/tutorial/introduction.html#strings
[1b]: https://docs.python.org/3/tutorial/introduction.html#lists
[1c]: https://docs.python.org/3/tutorial/datastructures.html#dictionaries
[1d]: https://docs.python.org/3/tutorial/datastructures.html#tuples-and-sequences
[2]: http://www.diveintopython.net/native_data_types/index.html
[3]: http://greenteapress.com/thinkpython2/html/index.html

**&copy; 2019: Gies College of Business at the University of Illinois.**

This notebook is released under the [Creative Commons license CC BY-NC-SA 4.0][ll]. Any reproduction, adaptation, distribution, dissemination or making available of this notebook for commercial use is not allowed unless authorized in writing by the copyright holder.

[ll]: https://creativecommons.org/licenses/by-nc-sa/4.0/legalcode